draft/XXXX-upload-system/README.md

```diff
@@ -0,0 +1,126 @@
+
+* If the upload takes too long, the connection times out (due to a current gunicorn setting)
+* If the connection is interrupted, the upload must start over from the beginning (very frustrating on large datasets)
+* There's no progress/indication to the user. This can be implemented on the _client_ by simply tracking the HTTP request progress, but as a matter of exposing this functionality through our API, it does not exist.
+
+## Proposals
+
+### Chunked Uploads
+Instead of dealing with a single request to upload data, this data could instead be sent in chunks, allowing for many smaller HTTP requests. This would allow the following things:
+* Retain a timeout on requests, while also allowing large datasets to be uploaded with ease.
+* Fault tolerance and upload pausing/resuming
+* Progress tracking
+
+This would be accomplished through the following additions.
+
+### Hidden Tables/Networks
+To allow for uploaded chunks to be stored somewhere, when starting an upload, the desired table/network would be created in a "hidden" mode. This would prevent the table from being returned by any table listing endpoints. This table should be assumed to be volatile, as data is still being uploaded to it in this state. Whether a table is hidden or not would be tracked by a `_meta` collection on each workspace. This would contain, among other things, the information of any hidden tables in that workspace. Each document in this collection will have a structure like so:
+
+```
+{
+  collection: id,
+  hidden: bool,
+}
+```
+
+The `collection` key would point to the ID of the collection to which this document refers.
+
+Other information could be added to this collection in subsequent RFCs, making this a general purpose collection metadata storage mechanism.
+
+#### Alternate Approach
+Instead of modifying the table/network listing endpoints to omit these "hidden" tables, these endpoints could instead be modified to return objects (dictionaries), which along with the table/networks names, would include if they are temporary (i.e. being actively uploaded to). The benefit to this would be that users wouldn't be able to accidentally create duplicate tables of the same name, as they could see all tables which may exist. However, this would allow for any user (who has access to the workspace) to see an in-progress table, which may contain garbage data, and is likely unvalidated. Furthermore, if an upload fails validation after the fact, it may once again disappear from the list of tables/networks, possibly causing more confusion.
+
+### System-wide Upload Collection
+To keep track of current uploads, a `_system` collection called `uploads` would be created, which would keep track of any ongoing uploads. This collection would store, among other things, the progress of the upload, the "hidden" table(s)/graph(s) to which it belongs, and the user who initiated the upload. The schema for each document in this collection would have a structure like so:
```

Comment:
When would the data here be cleaned out? When an upload fails or is successful?

Reply:
It would be good to decide on a policy for this. We could implement it with a cron-based cleanup script that runs at midnight and deletes upload objects older than, say, 24 hours or something. (I am not sure what the right "timeout" parameters are.)

Reply:
> We could implement it with a cron-based cleanup script that runs at midnight and deletes upload objects older than, say, 24 hours or something. (I am not sure what the right "timeout" parameters are.)

I'm weary of splitting this logic between the backend and the database server (i.e. more devops burden on us). I think once an upload has completed is one easy case to mark an upload as done, and remove it. As far as dealing with incomplete lingering uploads, I'm not sure what the best approach is. I have an idea centered around using celery, which also fits into other sections of this proposal. I'm going to discuss this more offline before bringing this idea here, as it's not fully formed yet.